eng/scripts/live-test-resource-cleanup.ps1

```diff
@@ -35,61 +47,226 @@ param (
+  $purgeableResources = @(Get-PurgeableResources)
+
+  Write-Host "Total Resource Groups To Delete: $($toDelete.Count)"
+  foreach ($rg in $toDelete)
+  {
+    $deleteAfter = GetTag $rg "DeleteAfter"
+    if ($Force -or $PSCmdlet.ShouldProcess("$($rg.ResourceGroupName) [DeleteAfter (UTC): $deleteAfter]", "Delete Group")) {
+      # Add purgeable resources that will be deleted with the resource group to the collection.
+      $purgeableResourcesFromRG = Get-PurgeableGroupResources $rg.ResourceGroupName
+
+      if ($purgeableResourcesFromRG) {
+        $purgeableResources += $purgeableResourcesFromRG
+        Write-Verbose "Found $($purgeableResourcesFromRG.Count) potentially purgeable resources in resource group $($rg.ResourceGroupName)"
+      }
+      Write-Verbose "Deleting group: $($rg.ResourceGroupName)"
+      Write-Verbose "  tags $($rg.Tags | ConvertTo-Json -Compress)"
+      # Write-Host ($rg | Remove-AzResourceGroup -Force -AsJob).Name
+    }
+  }
+
+  if (!$purgeableResources.Count) {
+    return
+  }
+  if ($Force -or $PSCmdlet.ShouldProcess("$($purgeableResources.VaultName)", "Delete Purgeable Resources")) {
+    # Purge all the purgeable resources and get a list of resources (as a collection) we need to follow-up on.
+    Write-Host "Attempting to purge $($purgeableResources.Count) resources."
+    $failedResources = @(Remove-PurgeableResources $purgeableResources -PassThru)
+    if ($failedResources) {
+      Write-Warning "Timed out deleting the following $($failedResources.Count) resources. Please file an IcM ticket per resource type."
+      $failedResources | Sort-Object AzsdkResourceType, AzsdkName | Format-Table -Property @{l='Type'; e={$_.AzsdkResourceType}}, @{l='Name'; e={$_.AzsdkName}}
+    }
+  }
+}
+
+function Login() {
+  if ($PSCmdlet.ParameterSetName -eq "Provisioner") {
+    Write-Verbose "Logging in with provisioner"
+    $provisionerSecret = ConvertTo-SecureString -String $ProvisionerApplicationSecret -AsPlainText -Force
+    $provisionerCredential = [System.Management.Automation.PSCredential]::new($ProvisionerApplicationId, $provisionerSecret)
+    Connect-AzAccount -Force -Tenant $TenantId -Credential $provisionerCredential -ServicePrincipal -Environment $Environment
+    Select-AzSubscription -Subscription $SubscriptionId -Confirm:$false
+  } elseif ($Login) {
+    Write-Verbose "Logging in with interactive user"
+    $cmd = "Connect-AzAccount"
+    if ($TenantId) {
+        $cmd += " -TenantId $TenantId"
+    }
+    if ($SubscriptionId) {
+        $cmd += " -SubscriptionId $SubscriptionId"
+    }
+    Invoke-Expression $cmd
+  } elseif (Get-AzContext) {
+    Write-Verbose "Using existing account"
+  } else {
+    $errMsg = 'User context not found. Please re-run script with "-Login" to login, ' +
+              'or run "Connect-AzAccount -UseDeviceAuthentication" if interactive login is not available.'
+    Write-Error $errMsg
+    exit 1
+  }
+}
+
+function Main() {
```

Comment:
Is there a reason to have a Main and function and just call it below?

Reply:
See https://github.com/Azure/azure-sdk-tools/pull/2847#discussion_r819872864

Reply:
That being the case, if the goal is to have helper functions not hiding the "real" script, I recommend defining `Main` first. You can use as-yet-undeclared functions in another function. They just have to be defined before calling any of them. Thus, you could define `Main`, then helper functions, then call main. IIRC, it was your goal to not have to list helper functions first.